File: app/src/main/java/com/limelight/StartExternalDisplayControlReceiver.java
package com.limelight;

import android.app.ActivityManager;
import android.app.ActivityOptions;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.Display;

import androidx.annotation.RequiresApi;

import com.limelight.utils.ExternalDisplayControlActivity;

public class StartExternalDisplayControlReceiver extends BroadcastReceiver {
    private static final long TIMEOUT_MS = 300;
    private static Handler handler = new Handler(Looper.getMainLooper());
    private static boolean isTimeoutActive = false;

    @RequiresApi(api = Build.VERSION_CODES.O)
    @Override
    public void onReceive(Context context, Intent intent) {
        requestFocusToGameActivity(true);
    }

    public static void requestFocusToExternalDisplayControl(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            Intent intentTouchpad = new Intent(context, ExternalDisplayControlActivity.class);
            intentTouchpad.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_SINGLE_TOP | Intent.FLAG_ACTIVITY_CLEAR_TOP);
            Bundle optionsDefault = ActivityOptions.makeBasic().setLaunchDisplayId(Display.DEFAULT_DISPLAY).toBundle();
            context.startActivity(intentTouchpad, optionsDefault);
        }
    }

    public static void requestFocusToGameActivity(boolean focusExternalDisplayControl) {
        if (isTimeoutActive) {
            return;
        }

        isTimeoutActive = true;

        if (Game.instance != null) {
            if (focusExternalDisplayControl) {
                requestFocusToExternalDisplayControl(Game.instance);
            }
            ActivityManager am = (ActivityManager) Game.instance.getSystemService(Context.ACTIVITY_SERVICE);
            if (am != null) {
                am.moveTaskToFront(Game.instance.getTaskId(), 0);
            }
        }

        handler.postDelayed(() -> isTimeoutActive = false, TIMEOUT_MS);
    }
}

File: app/src/main/java/com/papi/nova/StartExternalDisplayControlReceiver.kt
package com.papi.nova

import android.app.ActivityManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Looper

class StartExternalDisplayControlReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action != ACTION_START_EXTERNAL_DISPLAY_CONTROL) {
            return
        }

        requestFocusToGameActivity(true)
    }

    companion object {
        const val ACTION_START_EXTERNAL_DISPLAY_CONTROL =
            "com.papi.nova.action.START_EXTERNAL_DISPLAY_CONTROL"

        private const val TIMEOUT_MS = 300L
        private val handler = Handler(Looper.getMainLooper())
        private var isTimeoutActive = false

        @JvmStatic
        @Deprecated("Companion controls are now Game-owned; use requestFocusToGameActivity(true)")
        @Suppress("UNUSED_PARAMETER")
        fun requestFocusToExternalDisplayControl(context: Context) {
            requestFocusToGameActivity(true)
        }

        @JvmStatic
        @Deprecated("Companion controls are now Game-owned; use requestFocusToGameActivity(true)")
        @Suppress("UNUSED_PARAMETER")
        fun requestFocusToExternalDisplayControl(context: Context, streamDisplayId: Int) {
            requestFocusToGameActivity(true)
        }

        @JvmStatic
        fun requestFocusToGameActivity(showCompanionControls: Boolean) {
            if (isTimeoutActive) {
                return
            }

            isTimeoutActive = true
            val game = Game.instance
            val reopenRequestGeneration =
                if (showCompanionControls) game?.beginExplicitCompanionControlsReopen() else null
            if (game != null) {
                val activityManager = game.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager?
                activityManager?.moveTaskToFront(game.taskId, 0)
            }

            handler.postDelayed(
                {
                    if (
                        showCompanionControls &&
                        reopenRequestGeneration != null &&
                        game != null &&
                        Game.instance === game &&
                        !game.isFinishing
                    ) {
                        game.showCompanionControls(
                            explicitUserRequest = true,
                            requestGeneration = reopenRequestGeneration,
                        )
                    }
                    isTimeoutActive = false
                },
                TIMEOUT_MS,
            )
        }
    }
}
